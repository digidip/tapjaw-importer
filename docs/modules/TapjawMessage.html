<!DOCTYPE html><html class="default"><head><meta charSet="utf-8"/><meta http-equiv="x-ua-compatible" content="IE=edge"/><title>TapjawMessage | Tapjaw Importer API documentation - v2.0.0-rc7</title><meta name="description" content="Documentation for Tapjaw Importer API documentation - v2.0.0-rc7"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="stylesheet" href="../assets/style.css"/><link rel="stylesheet" href="../assets/highlight.css"/><script async src="../assets/search.js" id="search-script"></script></head><body><script>document.body.classList.add(localStorage.getItem("tsd-theme") || "os")</script><header><div class="tsd-page-toolbar"><div class="container"><div class="table-wrap"><div class="table-cell" id="tsd-search" data-base=".."><div class="field"><label for="tsd-search-field" class="tsd-widget search no-caption">Search</label><input type="text" id="tsd-search-field"/></div><ul class="results"><li class="state loading">Preparing search index...</li><li class="state failure">The search index is not available</li></ul><a href="../index.html" class="title">Tapjaw Importer API documentation - v2.0.0-rc7</a></div><div class="table-cell" id="tsd-widgets"><div id="tsd-filter"><a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a><div class="tsd-filter-group"><div class="tsd-select" id="tsd-filter-visibility"><span class="tsd-select-label">All</span><ul class="tsd-select-list"><li data-value="public">Public</li><li data-value="protected">Public/Protected</li><li data-value="private" class="selected">All</li></ul></div> <input type="checkbox" id="tsd-filter-inherited" checked/><label class="tsd-widget" for="tsd-filter-inherited">Inherited</label><input type="checkbox" id="tsd-filter-externals" checked/><label class="tsd-widget" for="tsd-filter-externals">Externals</label></div></div><a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a></div></div></div></div><div class="tsd-page-title"><div class="container"><ul class="tsd-breadcrumb"><li><a href="../index.html">Tapjaw Importer API documentation - v2.0.0-rc7</a></li><li><a href="TapjawMessage.html">TapjawMessage</a></li></ul><h1>Namespace TapjawMessage</h1></div></div></header><div class="container container-main"><div class="row"><div class="col-8 col-content"><section class="tsd-panel tsd-comment"><div class="tsd-comment tsd-typography"><dl class="tsd-comment-tags"><dt>namespace</dt><dd><p>TapjawMessage</p>
</dd></dl></div></section><section class="tsd-panel-group tsd-index-group"><h2>Index</h2><section class="tsd-panel tsd-index-panel"><div class="tsd-index-content"><section class="tsd-index-section "><h3>Classes</h3><ul class="tsd-index-list"><li class="tsd-kind-class tsd-parent-kind-namespace"><a href="../classes/TapjawMessage.TapjawMessage-1.html" class="tsd-kind-icon">Tapjaw<wbr/>Message</a></li></ul></section><section class="tsd-index-section "><h3>Type aliases</h3><ul class="tsd-index-list"><li class="tsd-kind-type-alias tsd-parent-kind-namespace"><a href="TapjawMessage.html#TapjawMessageDigest" class="tsd-kind-icon">Tapjaw<wbr/>Message<wbr/>Digest</a></li><li class="tsd-kind-type-alias tsd-parent-kind-namespace"><a href="TapjawMessage.html#TapjawPayload" class="tsd-kind-icon">Tapjaw<wbr/>Payload</a></li></ul></section><section class="tsd-index-section "><h3>Functions</h3><ul class="tsd-index-list"><li class="tsd-kind-function tsd-parent-kind-namespace tsd-has-type-parameter"><a href="TapjawMessage.html#sortObjectArrays" class="tsd-kind-icon">sort<wbr/>Object<wbr/>Arrays</a></li></ul></section></div></section></section><section class="tsd-panel-group tsd-member-group "><h2>Type aliases</h2><section class="tsd-panel tsd-member tsd-kind-type-alias tsd-parent-kind-namespace"><a id="TapjawMessageDigest" class="tsd-anchor"></a><h3>Tapjaw<wbr/>Message<wbr/>Digest</h3><div class="tsd-signature tsd-kind-icon">Tapjaw<wbr/>Message<wbr/>Digest<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">string</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/digidip/tapjaw-importer/blob/3ea39fc/src/messages/tapjaw-message.ts#L5">src/messages/tapjaw-message.ts:5</a></li></ul></aside></section><section class="tsd-panel tsd-member tsd-kind-type-alias tsd-parent-kind-namespace"><a id="TapjawPayload" class="tsd-anchor"></a><h3>Tapjaw<wbr/>Payload</h3><div class="tsd-signature tsd-kind-icon">Tapjaw<wbr/>Payload<span class="tsd-signature-symbol">:</span> <a href="_internal_.html#Record" class="tsd-signature-type" data-tsd-kind="Type alias">Record</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">, </span><span class="tsd-signature-type">unknown</span><span class="tsd-signature-symbol">&gt;</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/digidip/tapjaw-importer/blob/3ea39fc/src/messages/tapjaw-message.ts#L10">src/messages/tapjaw-message.ts:10</a></li></ul></aside><div class="tsd-comment tsd-typography"><div class="lead">
<p>Payload schema</p>
</div></div></section></section><section class="tsd-panel-group tsd-member-group "><h2>Functions</h2><section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace tsd-has-type-parameter"><a id="sortObjectArrays" class="tsd-anchor"></a><h3>sort<wbr/>Object<wbr/>Arrays</h3><ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace tsd-has-type-parameter"><li class="tsd-signature tsd-kind-icon">sort<wbr/>Object<wbr/>Arrays<span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type" data-tsd-kind="Type parameter">T</span><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol">(</span>payload<span class="tsd-signature-symbol">: </span><a href="_internal_.html#Record" class="tsd-signature-type" data-tsd-kind="Type alias">Record</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">, </span><span class="tsd-signature-type">unknown</span><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type" data-tsd-kind="Type parameter">T</span></li></ul><ul class="tsd-descriptions"><li class="tsd-description"><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/digidip/tapjaw-importer/blob/3ea39fc/src/support/sort-object-arrays.ts#L27">src/support/sort-object-arrays.ts:27</a></li></ul></aside><div class="tsd-comment tsd-typography"><div class="lead">
<p>This support method is used to prevent TapjawMessage
signatures incorrectly changing between API responses.</p>
</div><div><p>If a API payload contains properties which are arrays, in some
cases an API will provide the array in a different order between
requests which affects the signature of the TapjawMessage.</p>
<p>This function supports arrays containing objects or any scalar types.</p>
</div></div><h4 class="tsd-type-parameters-title">Type parameters</h4><ul class="tsd-type-parameters"><li><h4>T = <a href="_internal_.html#Record" class="tsd-signature-type" data-tsd-kind="Type alias">Record</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">, </span><span class="tsd-signature-type">unknown</span><span class="tsd-signature-symbol">&gt;</span></h4></li></ul><h4 class="tsd-parameters-title">Parameters</h4><ul class="tsd-parameters"><li><h5>payload: <a href="_internal_.html#Record" class="tsd-signature-type" data-tsd-kind="Type alias">Record</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">, </span><span class="tsd-signature-type">unknown</span><span class="tsd-signature-symbol">&gt;</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Payload</p>
</div></div></li></ul><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type" data-tsd-kind="Type parameter">T</span></h4></li></ul></section></section></div><div class="col-4 col-menu menu-sticky-wrap menu-highlight"><nav class="tsd-navigation primary"><ul><li class=""><a href="../index.html">Exports</a></li><li class=" tsd-kind-namespace"><a href="_internal_.html">&lt;internal&gt;</a></li><li class=" tsd-kind-namespace"><a href="TapjawAuthenticator.html">Tapjaw<wbr/>Authenticator</a></li><li class=" tsd-kind-namespace"><a href="TapjawCommand.html">Tapjaw<wbr/>Command</a></li><li class=" tsd-kind-namespace"><a href="TapjawConfig.html">Tapjaw<wbr/>Config</a></li><li class=" tsd-kind-namespace"><a href="TapjawConnector.html">Tapjaw<wbr/>Connector</a></li><li class=" tsd-kind-namespace"><a href="TapjawContract.html">Tapjaw<wbr/>Contract</a></li><li class=" tsd-kind-namespace"><a href="TapjawDate.html">Tapjaw<wbr/>Date</a></li><li class=" tsd-kind-namespace"><a href="TapjawError.html">Tapjaw<wbr/>Error</a></li><li class=" tsd-kind-namespace"><a href="TapjawIterator.html">Tapjaw<wbr/>Iterator</a></li><li class="current tsd-kind-namespace"><a href="TapjawMessage.html">Tapjaw<wbr/>Message</a></li><li class=" tsd-kind-namespace"><a href="TapjawMetadata.html">Tapjaw<wbr/>Metadata</a></li><li class=" tsd-kind-namespace"><a href="TapjawParser.html">Tapjaw<wbr/>Parser</a></li><li class=" tsd-kind-namespace"><a href="TapjawTypeguard.html">Tapjaw<wbr/>Typeguard</a></li></ul></nav><nav class="tsd-navigation secondary menu-sticky"><ul><li class="tsd-kind-class tsd-parent-kind-namespace"><a href="../classes/TapjawMessage.TapjawMessage-1.html" class="tsd-kind-icon">Tapjaw<wbr/>Message</a></li><li class="tsd-kind-type-alias tsd-parent-kind-namespace"><a href="TapjawMessage.html#TapjawMessageDigest" class="tsd-kind-icon">Tapjaw<wbr/>Message<wbr/>Digest</a></li><li class="tsd-kind-type-alias tsd-parent-kind-namespace"><a href="TapjawMessage.html#TapjawPayload" class="tsd-kind-icon">Tapjaw<wbr/>Payload</a></li><li class="tsd-kind-function tsd-parent-kind-namespace tsd-has-type-parameter"><a href="TapjawMessage.html#sortObjectArrays" class="tsd-kind-icon">sort<wbr/>Object<wbr/>Arrays</a></li></ul></nav></div></div></div><footer class="with-border-bottom"><div class="container"><h2>Settings</h2><p>Theme <select id="theme"><option value="os">OS</option><option value="light">Light</option><option value="dark">Dark</option></select></p></div></footer><div class="container tsd-generator"><p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p></div><div class="overlay"></div><script src="../assets/main.js"></script></body></html>