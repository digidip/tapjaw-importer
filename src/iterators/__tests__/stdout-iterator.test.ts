// Add custom configs
process.env.TAPJAW_MESSAGE_SHA256_SECRET = 'TestingSecet';

import StdoutIterator from '../stdout-iterator';
import TapjawMessage from '../../messages/tapjaw-message';
jest.mock('process');

describe('Make sure the iterator operates as expected', () => {
    test('should output 5 STDOUT calls', async () => {
        let stdoutOutput: (string | Uint8Array)[] = [];
        const spy = jest.spyOn(process.stdout, 'write').mockImplementation((msg: string | Uint8Array) => {
            stdoutOutput.push(msg);
            return true;
        });

        const iterator = new StdoutIterator(process.stdout);
        async function* callback(): AsyncGenerator<TapjawMessage> {
            yield new TapjawMessage('test1', { name: 'test1' }, new Date('2019-01-01'));
            yield new TapjawMessage('test2', { name: 'test2' }, new Date('2019-01-01'));
            yield new TapjawMessage('test3', { name: 'test3' }, new Date('2019-01-01'));
            yield new TapjawMessage('test4', { name: 'test4' }, new Date('2019-01-01'));
            yield new TapjawMessage('test5', { name: 'test5' }, new Date('2019-01-01'));
        }

        await iterator.run(callback);
        expect(spy).toBeCalledTimes(5);
        expect(stdoutOutput).toMatchSnapshot();
    });

    test('should throw an error when JSON error occurs', async () => {
        const spy = jest
            .spyOn(JSON, 'stringify')
            .mockImplementation(
                (
                    value: any,
                    replacer?: (string | number)[] | null | undefined,
                    space?: string | number | undefined
                ): string => {
                    return '';
                }
            );

        const iterator = new StdoutIterator(process.stdout);
        async function* callback(): AsyncGenerator<TapjawMessage> {
            yield new TapjawMessage('test1', { name: 'test1' }, new Date('2019-01-01'));
        }

        expect(iterator.run(callback))
            .rejects.toThrow('message could not be parsed into JSON.')
            .finally(() => {
                expect(spy).toBeCalled();
            });
    });

    test('should set pretty', async () => {
        const spy = jest
            .spyOn(JSON, 'stringify')
            .mockImplementation(
                (
                    value: any,
                    replacer?: (string | number)[] | null | undefined,
                    space?: string | number | undefined
                ): string => {
                    return '{}';
                }
            );

        const iterator = new StdoutIterator(process.stdout);

        const tapMessage = new TapjawMessage('test1', { name: 'test1' }, new Date('2019-01-01'));
        async function* callback(): AsyncGenerator<TapjawMessage> {
            yield tapMessage;
        }

        await iterator.run(callback);
        expect(spy).toBeCalledWith(tapMessage, null, undefined);

        iterator.setPretty(true);
        await iterator.run(callback);
        expect(spy).toBeCalledWith(tapMessage, null, 2);
    });
});
