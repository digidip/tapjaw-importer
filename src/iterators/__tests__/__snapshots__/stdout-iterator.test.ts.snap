// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make sure the iterator operates as expected should output 5 STDOUT calls 1`] = `
Array [
  "{\\"signature\\":\\"bef0fe89d19be845af6e0fd36b32a4ea8502ae49c91bd44f69532709a1ad8b6c\\",\\"sourceProviderName\\":\\"test1\\",\\"import_date\\":\\"2019-01-01T00:00:00.000Z\\",\\"payload\\":{\\"name\\":\\"test1\\"}}
",
  "{\\"signature\\":\\"72209b2b500e0921114d639de3831cf521f0a23afee11cf5d6e8daa931d83244\\",\\"sourceProviderName\\":\\"test2\\",\\"import_date\\":\\"2019-01-01T00:00:00.000Z\\",\\"payload\\":{\\"name\\":\\"test2\\"}}
",
  "{\\"signature\\":\\"a9ccd7b94cbaf85d6265639ffecbb0675c76a11a6a09d0ab31155c3dcc2229c4\\",\\"sourceProviderName\\":\\"test3\\",\\"import_date\\":\\"2019-01-01T00:00:00.000Z\\",\\"payload\\":{\\"name\\":\\"test3\\"}}
",
  "{\\"signature\\":\\"d1d4062e258e370078267819d7734e10b2176ce6453f94fae17febe587ec7f27\\",\\"sourceProviderName\\":\\"test4\\",\\"import_date\\":\\"2019-01-01T00:00:00.000Z\\",\\"payload\\":{\\"name\\":\\"test4\\"}}
",
  "{\\"signature\\":\\"a6db819c16ee9967d80c0e1ad808b4ac70ce75aed16c5b4bcecd13bc9f34b2a4\\",\\"sourceProviderName\\":\\"test5\\",\\"import_date\\":\\"2019-01-01T00:00:00.000Z\\",\\"payload\\":{\\"name\\":\\"test5\\"}}
",
]
`;
